src/main/scala/com/fulcrumgenomics/sopt/Sopt.scala

```diff
@@ -0,0 +1,158 @@
+  /** The result type when a single command is successfully parsed. */
+  case class CommandSuccess[A](command: A) extends Result[A,Nothing]
+  /** The result type when a command/sub-command pair are successfully parsed. */
+  case class SubcommandSuccess[A,B](command: A, subcommand: B) extends Result[A, B]
+  /** The result type when a parsing failure occurrs. */
+  case class Failure(usage: () => String) extends Result[Nothing,Nothing]
+
+  /**
+    * Represents the metadata about a command line program that may be consumed externally to generate
+    * documentation etc.
+    *
+    * @param name  the name of the program / command
+    * @param group the clp group name of the program
+    * @param hidden whether or not the program is marked as hidden
+    * @param description the description/documentation for the program in MarkDown format
+    * @param args the ordered [[Seq]] of arguments the program takes
+    */
+  case class ClpMetadata(name: String,
+                         group: String,
+                         hidden: Boolean,
+                         description: String,
+                         args: Seq[Arg]
+                        ) {
+    
+    /** Returns the description as plain text, line-wrapped to [[TerminalWidth]]. */
+    def descriptionAsText: String = markDownProcessor.toText(markDownProcessor.parse(description)).mkString("\n")
+
+    /** Returns the description as HTML. */
+    def descriptionAsHtml: String = markDownProcessor.toHtml(markDownProcessor.parse(description)).mkString("\n")
+  }
+
+  /**
+    * Represents information about an argument to a command line program.
+    *
+    * @param name the name of the argument as presented on the command line
+    * @param flag the optional flag character for the argument if it has one
+    * @param kind the name of the type of the argument (for collection arguments, the type in the collection)
+    * @param minValues the minimum number of values that must be specified
+    * @param maxValues the maximum number of values that may be specified
+    * @param defaultValues the seq of default values, as strings
+    * @param sensitive if true the argument is sensitive and values should not be re-displayed
+    * @param description the description of the argument
+    */
+  case class Arg(name: String,
+                 flag: Option[Char],
+                 kind: String,
+                 minValues: Int,
+                 maxValues: Int,
+                 defaultValues: Seq[String],
+                 sensitive: Boolean,
+                 description: String
+                ) {
+    /** Returns the description as plain text, line-wrapped to [[TerminalWidth]]. */
+    def descriptionAsText: String = markDownProcessor.toText(markDownProcessor.parse(description)).mkString("\n")
+
+    /** Returns the description as HTML. */
+    def descriptionAsHtml: String = markDownProcessor.toHtml(markDownProcessor.parse(description)).mkString("\n")
```

Comment:
Could mix in a trait instead of redefining:
```
trait Description {
  def description: String
  /** Returns the description as plain text, line-wrapped to [[TerminalWidth]]. *    
  def descriptionAsText: String = markDownProcessor.toText(markDownProcessor.parse(description)).mkString("\n")
  /** Returns the description as HTML. */
  def descriptionAsHtml: String = markDownProcessor.toHtml(markDownProcessor.parse(description)).mkString("\n")
}
```